### DISCUSSION TITLE
Decode CUE into Go interface types?

### DISCUSSION THREAD

**Discussion by heistp** [cue-lang/cue#1565] [Q&A]
Is there any way to specify the concrete Go type in CUE, so that a CUE value can be decoded into a Go interface type?

What I'm trying to do is, I'm using CUE for program configuration, and would like to allow configuration of a list of items each of which implements an interface, but I need some way to specify the concrete type in CUE so the decoder knows what underlying type to actually create and decode to.

With JSON, this is sometimes done using custom Marshal/Unmarshal methods ([example](https://gist.github.com/tkrajina/aec8d1b15b088c20f0df4afcd5f0c511)). This solution requires extra work and isn't all that clean, and I'm not sure if there's an equivalent in CUE.

I'm imagining that attributes could be used somehow, like:

```
{
   V: "val"
} @gotype(package.Struct)
```

I'm open to suggestions on how to do this a completely different way, but I would rather avoid using a type field and arbitrary properties, with error-prone boilerplate code in Go to deal with...

**Comment by heistp** [comment]
So far, I'm trying out using a union type, for example in CUE:

```
#S1: {
	F1: string
}

#S2: {
	F2: int
}

#Union: { S1: S1 } | { S2: S2 }

Setup: [ ...#Union ]

Setup: [
	{
		S1: {
			F1: "str"
		}
	},
	{
		S2: {
			F2: 3
		}
	},
]
```
And in Go:
```
package main

import (
	"fmt"
	"log"

	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

type Test struct {
	Setup []Union
}

type Union struct {
	S1 *S1
	S2 *S2
}

func (u Union) Print() {
	var p Printer
	switch {
	case u.S1 != nil:
		p = u.S1
	case u.S2 != nil:
		p = u.S2
	}
	p.Print()
}

type S1 struct {
	F1 string
}

func (s S1) Print() {
	fmt.Println(s.F1)
}

type S2 struct {
	F2 int
}

func (s S2) Print() {
	fmt.Println(s.F2)
}

type Printer interface {
	Print()
}

func main() {
	ctx := cuecontext.New()
	inst := load.Instances([]string{"test.cue"}, &load.Config{})

	for _, i := range inst {
		if i.Err != nil {
			log.Fatal(i.Err)
		}
		v := ctx.BuildInstance(i)
		if v.Err() != nil {
			log.Fatal(v.Err())
		}
		err := v.Validate()
		if err != nil {
			log.Fatal(err)
		}
		t := Test{}
		if err = v.Decode(&t); err != nil {
			log.Fatal(err)
		}
		var p Printer
		for _, s := range t.Setup {
			p = s
			p.Print()
		}
	}
}
```
This is one way to do it, but the CUE syntax is a bit cumbersome with structs that have single fields, and there's boilerplate to change on the Go side as new implementation types are added.

**Comment by mpvl** [comment]
One other way to do this would be to inspect the CUE `Value` before calling Decode to pick the relevant concrete Go type (in Go code). For instance, if all types are a result of unification with a definition, you can piece apart the `Value` for the occurrence of such a reference. Another approach would be to just try converting to a list of types and find the first one that succeeds.

You would need some support on the Go side to map keys to specific types, so you might as well do it in a way that doesn't require adding more support in CUE.
